Update for DTCalibration offline workflows

tobias-pook opened the pull request:
Running versions of the DT offline calibration was only available in private forks for some years now. This PR adds a working version again.

The changes include:

New python workflow system using crab3
Fixes to t0 determination
Cleanup and tweaks for segment selection.

serdweg committed:
Included the magentic field standard sequence

serdweg committed:
Update to the latest t0 changes

This commit includes changes which were done by several past
contributors, but never reached the official repo.

serdweg committed:
Included the missing consumes statements

tobias-pook committed:
DTCalibration Workflow update

This commit contains a rewrite of the DTcalibration workflow.
Main changes:

* All workflows inherit from one common DTCalibration object
* The CLI is dynamically build and each operations follows the structure
  dtCalibration [workflow] [workflow_mode] [command] using python
  subparsers
* Each worklflow is implemented in a single class.
* Each command needs to implement a function
* prepare_[workflow_mode]_[command]
    which prepares pset files for this command.
* Commonly used commands like submit, check, run are implemented in
  the parent class and may be ovewritten in the child classes if needed.
* Add new crab tools for crab3 integration
* automated proxy creation
* Delete obsolete files in workflows

Documentation about the updated workflow can be found in:
https://twiki.cern.ch/twiki/bin/viewauth/CMS/DTCalibrationTools

tobias-pook committed:
DTCalibration: Updates for new condDB usage

Imports of CondCore.DBCommon.ConDBSetup_cfi are deprecated and now
replaced by CondCore.CondDB.CondDB_cfi. The new object does not need the
auhentication mehtod parameter anymore and the connect parameter is
already set and is now set outside of the database objects.

ptraczyk committed:
cleanup residual fitting code

ptraczyk committed:
add residual history vs run number ntuple writer

ptraczyk committed:
small tweak in segment selection

ptraczyk committed:
add missing python file

ptraczyk committed:
add residual histogram range as configurable option

ptraczyk committed:
allow to run on RAW also when not MC

ptraczyk committed:
add default value for histogram range

ptraczyk committed:
update hlt path

ptraczyk committed:
bugfix in handling of histogram range definition through CLI

ptraczyk committed:
small bug fixes and fix logging

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
A new Pull Request was created by  @tobias-pook (Tobias) for master.

It involves the following packages:

CalibMuon/DTCalibration


@ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @lpernie can you please review it and eventually sign? Thanks.
@ghellwig, @mmusich, @tocheng this is something you requested to watch as well.
@davidlange6, @slava77 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20405/529

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/PR-20405/529/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/PR-20405/529/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly (this will soon be required for PRs to be considered)

arunhep commented:
@tobias-pook please implement the code-checker suggestions

tobias-pook committed:
Apply code-check suggestions

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
Pull request #20405 was updated. @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @lpernie can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20405/582

arunhep commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: d271ebf678263caa71c1041ee189da3584e05cc7


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/c72da2936d57a5c68e43b9182d7546b95a546365
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/22830/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/22830/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/22830/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
1001.0 step2
<pre>runTheMatrix-results/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5/step2_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5.log

</pre>1000.0 step2
<pre>runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step2_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log

</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/c72da2936d57a5c68e43b9182d7546b95a546365
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/22830/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/22830/git-merge-result

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

arunhep commented on CalibMuon/DTCalibration/python/ALCARECODtCalib_cff.py:
```diff
@@ -10,7 +10,7 @@
 #ALCARECODtCalibHLTFilter.HLTPaths = ['HLT_L1MuOpen*', 'HLT_L1Mu*']
 ALCARECODtCalibHLTFilter.throw = False ## dont throw on unknown path names
-ALCARECODtCalibHLTFilter.eventSetupPathsKey = 'DtCalib'
+ALCARECODtCalibHLTFilter.eventSetupPathsKey = 'MuAlcaDtCalibMu'
```
why this eventSetupPathsKey has been changed? What we have always is DtCalib. This is basically the reason for the failure of relvals because this key is part of a tag which contains all the keys associated to AlCaRecoTriggerBits. Please revert this change

tobias-pook replied:
I am waiting for a reply from the other contributors.

tobias-pook committed:
Revert unintended change in ALCARECODtCalib_cff

cmsbuild commented:
The code-checks are being triggered in jenkins.

tobias-pook commented:
It seems like this change was unintentionally introduced some time ago for testing by one of the contributors who already left the project. I reverted it now.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20405/726

cmsbuild commented:
Pull request #20405 was updated. @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @lpernie can you please check and sign again.

arunhep commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 27e70b6d9f82a3e7aeece3d3fa24562115eb424a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/23005/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/9f20abebebd79b197130b475a3a44ed3590e0f44
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/23005/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/23005/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/23005/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2646869
 - DQMHistoTests: Total failures: 206
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2646474
 - DQMHistoTests: Total skipped: 189
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 14 edm output root files, 26 DQM output files

tobias-pook commented:
@arunhep I see no pending or failed tests and comparisons anymore. Are there any more steps needed by our side to proceed with the PR?

arunhep commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar (and backports should be raised in the release meeting by the corresponding L2)

davidlange6 commented on CalibMuon/DTCalibration/interface/DTT0BaseCorrection.h:
```diff
@@ -4,6 +4,8 @@
 /** \class DTT0BaseCorrection
  *  Base class to define t0 corrections
  *
+ *  $Date: 2012/03/02 19:47:31 $
```
hi @tobias-pook - please remove these, they are obsolete (a sign that the merge of these developments has not accounted for evolution in CMSSW in the meantime?)

davidlange6 commented on CalibMuon/DTCalibration/plugins/DTResidualHistory.cc:
```diff
@@ -0,0 +1,308 @@
+     histoMapTH1F_.clear();
+     auto ch_it = dtGeom_->chambers().begin(); 	 
+     auto ch_end = dtGeom_->chambers().end(); 	 
+     for (; ch_it != ch_end; ++ch_it) { 	 
```
could these three loops be simplified into range based loops?

ptraczyk replied:
The comments on DTResidualHistory.cc apply also to other code in CalibMuon/DTCalibration/plugins - I guess this file was reviewed now because it was new. The loops above are copied over from DTResidualCalibration.cc, as is the histogram booking. std::stringstream is used in ~100 places across the whole directory. 
If these are to be fixed then I guess it makes sense to fix them everywhere, but its going to take a while

davidlange6 replied:
indeed - its a big job to update all legacy code - but that is not what is asked - rather we would like new code to be updated. Thanks

ptraczyk replied:
well I did do a slightly more extensive fix than just the new code, at least concerning loops and std::to_string.

davidlange6 commented on CalibMuon/DTCalibration/plugins/DTResidualHistory.cc:
```diff
@@ -0,0 +1,308 @@
+        }
+      }    
+
+    std::stringstream runStr; runStr << run;
```
please use std::to_string instead

davidlange6 commented on CalibMuon/DTCalibration/plugins/DTResidualHistory.cc:
```diff
@@ -0,0 +1,308 @@
+    if(!baseDir) baseDir = rootFile_->mkdir(("Run"+runStr.str()).c_str());
+    baseDir->cd();
+    histoResLs = new TH2F("histoResLs","Residuals vs Lumisection",100,0,10000,100,-1,1);
+    std::vector<TH2F*> histosTH2F;
```
please clean up pointers here and elsewhere- probably the best solution is just to make them smart pointers

davidlange6 replied:
to clarify - root does not "take care" of the memory for you. These need to be managed properly by your code.

ptraczyk replied:
I believe in this case smart pointers won't work, in fact putting them in makes the code segfault at the end of the execution. This is because with the call to "TH1AddDirectorySentry addDir" all the histograms are being managed by the TFile and the call to TFile::Close() at the end deletes them. Then they try to delete themselves again when DTResidualHistory is destroyed

davidlange6 commented on CalibMuon/DTCalibration/plugins/DTResidualHistory.cc:
```diff
@@ -0,0 +1,308 @@
+  }
+}
+
+float DTResidualHistory::segmentToWireDistance(const DTRecHit1D& recHit1D, const DTRecSegment4D& segment){
```
it looks like this function already exists in CMSSW - please make a helper class to share code.
https://github.com/cms-sw/cmssw/blob/09c3fce6626f70fd04223e7dacebf0b485f73f54/CalibMuon/DTCalibration/plugins/DTResidualCalibration.h#L43

davidlange6 commented on CalibMuon/DTCalibration/plugins/DTResidualHistory.cc:
```diff
@@ -0,0 +1,308 @@
+  cout << "[DTResidualHistory] Booking histos for SL: " << slId << endl;
+
+  // Compose the chamber name
+  std::stringstream runStr; runStr << run;
```
use std::to_string for all of these

davidlange6 commented on CalibMuon/DTCalibration/plugins/SealModule.cc:
```diff
@@ -48,7 +50,7 @@ DEFINE_FWK_MODULE(DTTTrigWriter);
 DEFINE_FWK_MODULE(DTTTrigCorrectionFirst);
 DEFINE_FWK_MODULE(DTT0Calibration);
-//DEFINE_FWK_MODULE(DTT0CalibrationNew);
+DEFINE_FWK_MODULE(DTT0CalibrationNew);
```
what is the plan for "new" becoming standard? Can we remove the old one?

tobias-pook replied:
At the moment the "new" DTT0 calibration is mostly used but the old method is still used sometimes for crosschecks. We are working on a rewrite of the DTT0 calibration code and will replace this class in a future PR.

davidlange6 replied:
@tobias-pook - please rename as V2 or some other descriptive identifier rather than "new" which is rather meaningless after a while - these replacements sometimes seem to take much longer than originally hoped..

davidlange6 commented on CalibMuon/DTCalibration/scripts/lcg-lsR.sh:
```diff
@@ -0,0 +1,62 @@
+#!/bin/bash
```
please give this a dt specific name (or perhaps its already provided by an wlcg environment and should not be in CMSSW?)

davidlange6 commented:
-1
pending comments to be addressed

tobias-pook committed:
Remove obsolete svn tags

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20405/865

cmsbuild commented:
Pull request #20405 was updated. @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @lpernie can you please check and sign again.

tobias-pook committed:
Remove obsolete lcg script for crab output on lxplus

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20405/867

cmsbuild commented:
Pull request #20405 was updated. @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @lpernie can you please check and sign again.

ptraczyk committed:
clean up histogram naming code and loops

cmsbuild commented:
The code-checks are being triggered in jenkins.

ptraczyk committed:
remove unintentional file

cmsbuild commented:
The code-checks are being triggered in jenkins.

ptraczyk committed:
fixes from code-checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20405/915

cmsbuild commented:
Pull request #20405 was updated. @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @lpernie can you please check and sign again.

ptraczyk committed:
remove unused method in DTResidualHistory

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20405/920

cmsbuild commented:
Pull request #20405 was updated. @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @lpernie can you please check and sign again.

ptraczyk commented:
I addressed the comments about DTResidualHistory.cc (and also made fixes in other files where I found similar issues) ... except the one about smart pointers - are we really supposed to use smart pointers to TH1F? I thought ROOT takes care of memory management for these..
(in fact I tried just to check, but I can't get it to compile even, though I might be doing something wrong)

arunhep commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 97a9c170b0586daf37592c41cc6b11a82943102b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/23218/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/630acaf4d413c283cf6a9ae3be6037f071b994cb
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/23218/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/23218/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/23218/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2691683
 - DQMHistoTests: Total failures: 221
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2691273
 - DQMHistoTests: Total skipped: 189
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 15 edm output root files, 26 DQM output files

tobias-pook commented:
Any updates ? We have addressed all comment.

arunhep commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar (and backports should be raised in the release meeting by the corresponding L2)

ptraczyk committed:
remove DTResidualHistory

cmsbuild commented:
The code-checks are being triggered in jenkins.

ptraczyk commented:
I removed DTResidualHistory, this can go in later in another PR

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20405/1376

cmsbuild commented:
Pull request #20405 was updated. @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @lpernie can you please check and sign again.

arunhep commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 4848c76feeae895d1d495b202fd6c518487ecd0d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/23776/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/42f293a55d70ffcb847798f20fd9102259c7f8e8
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/23776/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/23776/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/23776/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2767139
 - DQMHistoTests: Total failures: 106
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2766862
 - DQMHistoTests: Total skipped: 171
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 10 edm output root files, 26 DQM output files

tobias-pook committed:
Rename DTT0Calibration plugins

Renamed DTT0Calibration to DTT0CalibrationRMS as it represents a
simple baseline based on the peak RMS.
Accordingly the more sophisticated DTT0CalibrationNew plugin is
now renamed to DTT0Calibration

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20405/1469

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/PR-20405/1469/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/PR-20405/1469/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

tobias-pook committed:
Apply code check patch and adjust one logging line

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20405/1470

cmsbuild commented:
Pull request #20405 was updated. @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @lpernie can you please check and sign again.

arunhep commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 37abaa8f2d51e4e6b37755c387479038eae71f3c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/23937/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/54ee681cd4d890494361d5fb62c374f202c241ce
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/23937/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/23937/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20405/23937/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2815041
 - DQMHistoTests: Total failures: 113
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2814757
 - DQMHistoTests: Total skipped: 171
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 10 edm output root files, 26 DQM output files

tobias-pook commented:
Any updates about the pull request ?

ptraczyk commented:
Hi, could we go ahead with this PR? We believe we've addressed all the comments

lpernie commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar (and backports should be raised in the release meeting by the corresponding L2)

ptraczyk commented:
Hi, we seem to be stuck with this PR again..

lpernie commented:
@davidlange6 Could we merge this PR?
It was also in the last ORP.

lpernie commented:
Hi @davidlange6 sorry to insist, but what is holding this PR?

davidlange6 commented:
10k lines of changes?

> On Dec 4, 2017, at 5:41 PM, Luca Pernie <notifications@github.com> wrote:
> 
> Hi @davidlange6 sorry to insist, but what is holding this PR?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

lpernie commented:
I asked because DT people are waiting for it, and asked me for a time estimation.
I will tell them it needs more time due to the number of files modified.

davidlange6 commented:
+1
sorry for the slow review
